### DISCUSSION TITLE
Issues overloading set() function on custom ValueConsumer for a display

### DISCUSSION THREAD

**Discussion by caballero03** [SignalK/SensESP#790] [Q&A] [answered]
I have been experimenting with SensESP v3.0 and mostly having great success. However, as I try to do more advanced things I have run into a snag. What I am trying to do (maybe mistakenly) is have a custom ValueConsumer class that when provided a float value will display the numeric value or graphical representation of the value on the display and when provided with an integer value, it will apply that to the LCD brightness.

I have all of this working with two separate classes; one for displaying values and one for backlight brightness control. My initial design was to combine these in a way that might make it possible to change the display color to red when the ambient light level is low in addition to dimming the LCD backlight.

Can anyone shed some light on what I might need to do here? TIA

Here are some snippets of my code and error messages that I am getting when trying to make a custom polymorphic ValueConsumer class:

```C++
  // MyDisplay.h class definition
  
  class MyDisplay  : public FloatConsumer, public FloatProducer//, public IntConsumer <--- Cannot get IntConsumer to work without breaking things
                     {
	public:
		// Constructors
		MyDisplay( ...);

		MyDisplay( ...);

		virtual void set(const float& new_value) override;
		virtual void set(const int& new_value); // <-- Can't override without IntConsumer (of course)
		
	... other methods ...
  }




  // Usage of class in main.cpp and errors encountered

  const char* DBT_sk_path = "environment.depth.belowTransducer";

  /////////////////////////////////////////////
  // Depth below transducer
  auto* depthListener = new FloatSKListener(DBT_sk_path);

  depthListener
    ->connect_to(new MyDisplay(&tft, ROW_1, COL_3, "m", FONT_2)) // <--- connect_to() fails with "no instance of overloaded function "sensesp::SKValueListener<T>::connect_to [with T=float]" matches the argument list" when IntConsumer is added to class header
    // Convert meters to feet
    ->connect_to(new Linear(METERS_TO_FEET, 0.0))
    ->connect_to(new MyDisplay(&tft, ROW_5, COL_3, "ft", FONT_2))
    ->connect_to(new MyDisplay(&tft, DisplayType::VBAR, ROW_13, COL_10, 100, "ft", FONT_2));
```

**Comment by mairas** [comment, marked as answer]
Having a single class inheriting from multiple consumer types was a pattern used in previous SensESP versions, but I removed support for it during the version 3 development. I believe some technical reasons (that I fail to remember) were the breaking point, but I also didn't like the pattern much. It wasn't possible to create multiple consumers of the same type, and the usage of such multi-consumers could be confusing.

However, there's a better pattern available. In `MyDisplay`, define public `LambdaConsumer` members for the inputs you need. For example, for brightness, you would have:

```
...
 public:
  LambdaConsumer<int> brightness_{[this](int value) {
   this->set_backlight(value);   
  }};
```

From a class user POV (even if you were the same person 😁), the intent is much clearer if you can do:

```
brightness_producer->connect_to(my_display->brightness_);
```

rather than having two different producers connect to what seems like the same consumer.

Have a look at the [`SystemStatusController`](https://github.com/SignalK/SensESP/blob/main/src/sensesp/controllers/system_status_controller.h) implementation for an example - although the public getters are an unnecessary distraction, IMO. (I can say that because I've written that code. 😁)

**Reply by caballero03** [reply]
Thank you very much for your prompt reply and excellent advice. I do think the new pattern makes much more intuitive sense as you had mentioned. 

If anyone wants to see what I finally came up with, I have uploaded my code for the display on Github. It is very much a works-in-progress and may serve to supply some nice giggles or even a belly laugh or two. Enjoy. :-P
https://github.com/caballero03/SignalK-SensESP-CYD-Display

**Reply by mairas** [reply]
Cool stuff!

Also have a look at LVGL - in the long run, it might be easier to use ready-made UI elements and controls. LVGL is a very popular graphics library and there are tutorials specifically for using it with the CYD. If you have trouble figuring out how to adapt LVGL examples to SensESP, just ask and we'll figure it out.

**Reply by caballero03** [reply]
Thank you for your kind words of encouragement. I started this project almost by accident and it came together so quickly that I just kept going with it. I have yet to take time to plan much of anything out. This CYD project, when "finished", should be a helpful display for my boat systems' monitoring while aboard. 

I agree that using LVGL with its ready-made UI elements would be easier in the long run. I was actually thinking about that initially but need I to ease into such an undertaking. I will be having much more free time this winter (seasonal job) and may try to tackle something like LVGL soon.

Currently, I am not having much love when trying to compile the LVGL examples for the CYD that I have found. It is probably due to me being new to using and configuring VSCode/PlatformIO projects. I just get a bunch of perplexing dependency missing errors. I'll give it another go very soon.

It'd be great if I could pick your brain for the nitty-gritty of getting LVGL integrated with SensESP. Thank you for your offer to help out. I have been very impressed with the SensESP system as I work with it more and more. Well done, sir!

**Comment by mairas** [comment]
I'm changing this issue into a discussion in hopes that it can be found easier in case someone else is struggling with the same thing.
